### ISSUE TITLE
Proposal: Special case behavior for SA1501 when SA1503 is suppressed

### ISSUE THREAD

**Issue by sharwell** [DotNetAnalyzers/StyleCopAnalyzers#1175]
## Background

Currently, code like the following violates SA1503:

```csharp
if (x) return;
```

To fix this violation, the user adds braces. The following "simplest" addition of braces violates SA1501:

```csharp
if (x) { return; }
```

To resolve both violations, the user has to write the statement as follows:

```csharp
if (x)
{
    return;
}
```

## Problem

Teams that allow simple statements to be written without braces may disable SA1503. When SA1503 is disabled, the original snippet above does not report any violation at all.

## Proposal

When SA1503 is disabled, I propose modifying the behavior of SA1501 to disallow the following statement:

```csharp
if (x) return;
```

To resolve the issue, the code fix should move the return statement to the next line:

```csharp
if (x)
    return;
```

:memo: The documentation for SA1501 would need to be updated. The title and message format already make sense with this proposal, but the Cause section of the documentation specifically mentions braces as the item of interest.

**Comment by JohanLarsson** [comment]
Single line is often used for `if(x == null) throw new ArgumentNullException("x")`
Dunno if throwing should be special cased or if it even requires a special case.

With return on a separate line indentation is pretty important. Perhaps the next line after the return should also be checked so it is not indented, likely error if so.

Side note:
Is there something that catches:
```
if(x); // ends with semicolon
    return 2.3;
```
I have had to debug ^ a couple of times :)

**Comment by sharwell** [comment]
@JohanLarsson your example is one of the cases I'm specifically trying to catch. If you want single-line argument validation, use calls instead of block structure. For example:

```csharp
Requires.NotNull(nameof(x));
```

The second example should be caught by SA1106 (CodeMustNotContainEmptyStatements).

**Comment by vweijsters** [comment]
:+1: I like this proposal. For me this matches nicely with the spirit of the original StyleCop.

(I would personally never disable SA1503, but that's a different subject of discussion :wink:)
